common_examples/tcp_client/main/tcp_client.c

```diff
@@ -0,0 +1,102 @@
+    ESP_LOGI(TAG, "ETHGW:" IPSTR, IP2STR(&ip_info->gw));
+    ESP_LOGI(TAG, "~~~~~~~~~~~");
+    xSemaphoreGive(x_got_ip_semaphore);
+}
+
+void app_main(void)
+{
+    // Create default event loop that running in background
+    ESP_ERROR_CHECK(esp_event_loop_create_default());
+    // Initialize semaphore
+    x_got_ip_semaphore = xSemaphoreCreateBinary();
+    // Initialize Ethernet driver
+    uint8_t eth_port_cnt = 0;
+    esp_eth_handle_t *eth_handles;
+    char if_key_str[10];
+    char if_desc_str[10];
+    ESP_ERROR_CHECK(ethernet_init_all(&eth_handles, &eth_port_cnt));
+    esp_netif_init();
+    // Register user defined event handers
+    ESP_ERROR_CHECK(esp_event_handler_register(IP_EVENT, IP_EVENT_ETH_GOT_IP, &got_ip_event_handler, NULL));
+    // Create instance(s) of esp-netif for Ethernet(s)
+    if (eth_port_cnt == 1) {
+        // Use ESP_NETIF_DEFAULT_ETH when just one Ethernet interface is used and you don't need to modify
+        // default esp-netif configuration parameters.
+        esp_netif_config_t cfg = ESP_NETIF_DEFAULT_ETH();
+        esp_netif_t *eth_netif = esp_netif_new(&cfg);
+        // Attach Ethernet driver to TCP/IP stack
+        ESP_ERROR_CHECK(esp_netif_attach(eth_netif, esp_eth_new_netif_glue(eth_handles[0])));
+    } else {
+        // Use ESP_NETIF_INHERENT_DEFAULT_ETH when multiple Ethernet interfaces are used and so you need to modify
+        // esp-netif configuration parameters for each interface (name, priority, etc.).
+        esp_netif_inherent_config_t esp_netif_config = ESP_NETIF_INHERENT_DEFAULT_ETH();
+        esp_netif_config_t cfg_spi = {
+            .base = &esp_netif_config,
+            .stack = ESP_NETIF_NETSTACK_DEFAULT_ETH
+        };
+
+        for (int i = 0; i < eth_port_cnt; i++) {
+            sprintf(if_key_str, "ETH_%d", i);
+            sprintf(if_desc_str, "eth%d", i);
+            esp_netif_config.if_key = if_key_str;
+            esp_netif_config.if_desc = if_desc_str;
+            esp_netif_config.route_prio -= i * 5;
+            esp_netif_t *eth_netif = esp_netif_new(&cfg_spi);
+
+            // Attach Ethernet driver to TCP/IP stack
+            ESP_ERROR_CHECK(esp_netif_attach(eth_netif, esp_eth_new_netif_glue(eth_handles[i])));
+            esp_eth_start(eth_handles[i]);
+        }
+    }
+    // Wait until IP address is assigned to this device
+    xSemaphoreTake(x_got_ip_semaphore, portMAX_DELAY);
+    ESP_LOGI(TAG, "TCP client has started, waiting for the server to accept a connection.");
+    int client_fd;
+    struct sockaddr_in server;
+    char rxbuffer[SOCKET_MAX_LENGTH] = {0};
+    char txbuffer[SOCKET_MAX_LENGTH] = {0};
+    client_fd = socket(AF_INET, SOCK_STREAM, 0);
+    server.sin_family = AF_INET;
+    server.sin_port = htons(SOCKET_PORT);
+    server.sin_addr.s_addr = inet_addr(CONFIG_EXAMPLE_SERVER_IP_ADDRESS);
+    ESP_ERROR_CHECK(connect(client_fd, (struct sockaddr *)&server, sizeof(struct sockaddr)));
```

Comment:
I'd suggest that we log an error and try again on connection failure. The reason is that users would to complain that the program "crashes and resets permanently" if the server address was not available (and I agree that it's hard to read error messages)

Another point `connect()` doesn't return `esp_err_t`, so better to compare to `0` and `-1`. than to `ESP_OK`

Reply:
Now making multiple tries while logging an error each time it fails. Should delay be used between them, or any limit on tries be placed.

Reply:
You can just check the error like this:

https://github.com/espressif/esp-idf/blob/3b8741b172dc951e18509698dee938304bcf1523/examples/protocols/sockets/tcp_client_multi_net/main/tcp_client_multiple.c#L83-L85